src/visualization.cpp

```diff
@@ -50,101 +49,130 @@ void Visualization::publishPlanPoint(const std_msgs::Header & header, const tf2:
-  mkCollisionHull.pose.orientation.w = 1.0;
-  mkCollisionHull.id = __COUNTER__;
-  mkCollisionHull.type = visualization_msgs::Marker::LINE_STRIP;
-  mkCollisionHull.scale.x = 0.2;
-  mkCollisionHull.color.r = 1.0;
-  mkCollisionHull.color.a = 0.3;
-  mkCollisionHull.points = hull;
-
-  visualization_msgs::Marker mkCollisionIndicator;
-  mkCollisionIndicator.header.frame_id = frame_id;
-  mkCollisionIndicator.header.stamp = ros::Time::now();
-  mkCollisionIndicator.ns = "Collision object";
-  mkCollisionIndicator.pose.orientation.w = 1.0;
-  mkCollisionIndicator.id = __COUNTER__;
-  mkCollisionIndicator.type = visualization_msgs::Marker::CYLINDER;
-  mkCollisionIndicator.scale.x = 0.5;
-  mkCollisionIndicator.scale.y = 0.5;
-  mkCollisionIndicator.color.r = 1.0;
-  mkCollisionIndicator.color.a = 0.0;
-
-  visualization_msgs::MarkerArray mkCollision;
-  mkCollisionIndicator.scale.z = cost / 255.0;
-  mkCollisionIndicator.color.a = cost / 255.0;
-  mkCollisionIndicator.pose.position = point;
-  mkCollisionIndicator.pose.position.z = mkCollisionIndicator.scale.z * 0.5;
-  if (mkCollisionIndicator.scale.z > std::numeric_limits<float>::epsilon()) {
-    mkCollisionIndicator.action = visualization_msgs::Marker::ADD;
+  visualization_msgs::Marker marker;
+  marker.header = header;
+  marker.ns = "Collision object";
+  marker.pose.orientation.w = 1.0;
+  marker.id = __COUNTER__;
+  marker.type = visualization_msgs::Marker::CYLINDER;
+  marker.scale.x = 0.5;
+  marker.scale.y = 0.5;
+  marker.color.r = 1.0;
+  marker.color.a = 0.0;
+  marker.scale.z = cost / 255.0;
+  marker.color.a = cost / 255.0;
+  tf2::toMsg(point, marker.pose.position);
+  marker.pose.position.z = marker.scale.z * 0.5;
+  if (marker.scale.z > std::numeric_limits<float>::epsilon()) {
+    marker.action = visualization_msgs::Marker::ADD;
   } else {
-    mkCollisionIndicator.action = visualization_msgs::Marker::DELETE;
+    marker.action = visualization_msgs::Marker::DELETE;
   }
 
-  mkCollision.markers.push_back(mkCollisionIndicator);
+  marker_pub_.publish(marker);
+}
 
-  mkCollision.markers.push_back(mkCollisionFootprint);
-  mkCollision.markers.push_back(mkCollisionHull);
-  if (!mkSteps.points.empty()) {
-    mkCollision.markers.push_back(mkSteps);
-  }
-  if (!mkPosesOnPath.points.empty()) {
-    mkCollision.markers.push_back(mkPosesOnPath);
-  }
+void Visualization::publishExtrapolatedPoses(
+  const std_msgs::Header & header, const std::vector<tf2::Vector3> & steps)
+{
+  visualization_msgs::Marker marker;
+  marker.header = header;
+  marker.ns = "extrapolated poses";
+  marker.action = visualization_msgs::Marker::ADD;
+  marker.pose.orientation.w = 1.0;
+  marker.id = __COUNTER__;
+  marker.type = visualization_msgs::Marker::POINTS;
+  marker.scale.x = 0.5;
+  marker.scale.y = 0.5;
+  marker.color.r = 1.0;
+  marker.color.g = 0.5;
+  marker.color.a = 1.0;
+
+  std::transform(
+    steps.begin(), steps.end(), std::back_inserter(marker.points), [](const auto & step) {
```

Comment:
You use the same lambda multiple times, consider creating a simple function (in an anonymous namespace at the top of the source file) and pass that to `std::transform()`.

Reply:
And/or create a function for the entire container conversion:

```cpp
auto convert_points(const std::vector<tf2::Vector3>& points)
{
  ...
  std::transform(...);
  ...
}
```

Reply:
Fixed, I've create a utility function for this conversion